iree/compiler/Dialect/Flow/IR/FlowOps.cpp

```diff
@@ -565,37 +323,76 @@ bool DispatchWorkgroupsOp::canClosureContainOp(Operation *op) {
   return canDispatchRegionContainOp(op);
 }
 
-bool DispatchWorkgroupsOp::isOutputReadWithinRegion(unsigned resultIndex) {
+// Refines the tensor access from what is declared on |type| based on actual
+// usage. We expect that the access was set correctly to begin with but today
+// we sometimes specify things too wide.
+static TensorAccess refineTensorAccess(Value value, DispatchTensorType type) {
```

Comment:
Maybe we could add a verifier to check that a readwrite has a read and a write. This will help catch the places where dispatch region creation is doing the wrong thing?

Reply:
I'm going to try to use memory effects for this in a future change - then a canonicalization step will take care of narrowing this and changing the type instead of having all the usage sites query it like this.